### PR TITLE
Adding ZeroShotImageClassificationPipeline

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

- Based on CLIP

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@suraj-patil @LysandreJik

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Narsil** [comment]
Pinging @patil-suraj @LysandreJik

**Comment by Narsil** [comment]
Friendly ping @LysandreJik  @patil-suraj

**Review by LysandreJik — COMMENTED** [review]
Thanks for the ping! From a quick look it looks good to me - but I'll let @patil-suraj check out the actual image classification code.

Suraj is hard at work on the Flax/Jax sprint so the review will probably come once it has calmed down a bit :)

**Comment by julien-c** [comment]
unstale?

**Review by patil-suraj — COMMENTED** [review]
This is looking good, thanks a lot for adding this! Left a few comments.

Could you explain a bit how batching is handled ?

**Comment by Narsil** [comment]
> This is looking good, thanks a lot for adding this! Left a few comments.

Thanks for this, many places where doc/code was out of sync/bad copy paste here.

> Could you explain a bit how batching is handled ?

https://github.com/huggingface/transformers/pull/14225

This should contain more information on how it's handled internally. The pseudo code and images try to convey how it's done.
Tell me how this could be improved, it should belong in the doc actually.

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+from typing import List, Union
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_image_classification.py:1]: Really ? I never understood if/why such a thing was ever needed.

File: src/transformers/pipelines/base.py [hunk 2]
```diff
@@ -105,7 +105,8 @@ def _pad(items, key, padding_value, padding_side):
 
 
 def pad_collate_fn(tokenizer, feature_extractor):
-    padding_side = "right"
+    t_padding_side = None
+    f_padding_side = None
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/base.py:109]: Done.

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+if is_torch_available():
+    import torch
+
+logger = logging.get_logger(__name__)
+
+
+@add_end_docstrings(PIPELINE_INIT_ARGS)
+class ZeroShotImageClassificationPipeline(ChunkPipeline):
+    """
+    Image classification pipeline using any `AutoModelForZeroShotImageClassification`. This pipeline predicts the class
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_image_classification.py:22]: Done.

[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_image_classification.py:22]: Done.

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    of an image.
+
+    This image classification pipeline can currently be loaded from [`pipeline`] using the following task identifier:
+    `"image-classification"`.
+
+    See the list of available models on
+    [huggingface.co/models](https://huggingface.co/models?filter=image-classification).
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_image_classification.py:29]: I modified the comment right now. We need to update things on the hub once we have the widgets running.

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+                The pipeline accepts either a single image or a batch of images, which must then be passed as a string.
+                Images in a batch must all be in the same format: all as http links, all as local paths, or all as PIL
+                images.
+            candidate_labels (`List[str]`):
+                The candidate labels for this image
+            hypothesis_template (`str`, *optional*, defaults to `"This is a photo of a {}"`):
+                The sentence used in cunjunction with *candidate_labels* to attempt the image classification by
+                replacing the placeholder with the candidate_labels. Then likelihood is estimated by using
+                likelihood_per_image
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_image_classification.py:62]: Yes.

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    def preprocess(self, image, candidate_labels=None, hypothesis_template="This is a photo of {}."):
+        n = len(candidate_labels)
+        for i, candidate_label in enumerate(candidate_labels):
+            image = load_image(image)
+            images = self.feature_extractor(images=[image], return_tensors="pt")
+            sequence = hypothesis_template.format(candidate_label)
+            inputs = self.tokenizer(sequence, return_tensors="pt")
+            inputs["pixel_values"] = images.pixel_values
+            yield {"is_last": i == n - 1, "candidate_label": candidate_label, **inputs}
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_image_classification.py:96]: Yes

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    def _sanitize_parameters(self, **kwargs):
+        preprocess_params = {}
+        if "candidate_labels" in kwargs:
+            preprocess_params["candidate_labels"] = kwargs["candidate_labels"]
+        if "hypothesis_template" in kwargs:
+            preprocess_params["hypothesis_template"] = kwargs["hypothesis_template"]
+
+        postprocess_params = {}
+        if "multi_label" in kwargs:
+            postprocess_params["multi_label"] = kwargs["multi_label"]
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_image_classification.py:85]: Indeed.

File: tests/test_pipelines_zero_shot_image_classification.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    #     # Batching
+    #     outputs = pipe([image] * 3, batch_size=2, candidate_labels=["A", "B"])
+
+    @require_tf
+    def test_small_model_tf(self):
+        self.skipTest("Not implemented in Tensorflow")
+
+    @require_torch
+    def test_small_model_pt(self):
+        speech_recognizer = pipeline(
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tests/test_pipelines_zero_shot_image_classification.py:80]: Thanks.

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        return preprocess_params, {}, postprocess_params
+
+    def preprocess(self, image, candidate_labels=None, hypothesis_template="This is a photo of {}."):
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_image_classification.py:80]: It needs to be sent as a parameter since that was how it was done in  `zero-shot-classification`.
(debatable choice, but it does enable `pipe = pipeline(..., candidate_labels["hot-dog", "pizza"]), pipe(image_dataset)`)

Parameters can't be `[]` by default (first no mutable default arguments, second it does not make sense to run the pipeline on an empty set of  candidates.

2 solutions to improve the error when user forgets candidate_labels: (current it will be something along the  llines that None is not iterable)

- add `if candidate_labels is None: raise Exception("BETTER EXCEPTION MESSAGE")`. simple, I like it.
- change signature to `preprocess(self, image, *, candidate_labels,...)`. I need to check, but I think that will give a nice error message directly sent by Python.

File: src/transformers/pipelines/__init__.py [hunk 10]
```diff
@@ -239,6 +240,13 @@
         },
         "type": "text",
     },
+    "zero-shot-image-classification": {
+        "impl": ZeroShotImageClassificationPipeline,
+        "tf": (),
```
Comment (patil-suraj) [inline src/transformers/pipelines/__init__.py:245]: CLIP is also available in TF.

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+                The sentence used in cunjunction with *candidate_labels* to attempt the image classification by
+                replacing the placeholder with the candidate_labels. Then likelihood is estimated by using
+                logits_per_image
+
+        Return:
+            A list of dictionaries containing result, one dictionnary per proposed label. The dictionaries contain the
+            following keys:
+
+            - **label** (`str`) -- The label identified by the model. It is one of the suggested `candidate_label`.
+            - **score** (`int`) -- The score attributed by the model for that label (between 0 and 1).
```
Comment (patil-suraj) [inline src/transformers/pipelines/zero_shot_image_classification.py:67]:
```suggestion
            - **score** (`float`) -- The score attributed by the model for that label (between 0 and 1).
```

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+    See the list of available models on
+    [huggingface.co/models](https://huggingface.co/models?filter=zer-shot-image-classification).
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        if self.framework != "pt":
+            raise ValueError(f"The {self.__class__} is only available in PyTorch.")
```
Comment (patil-suraj) [inline src/transformers/pipelines/zero_shot_image_classification.py:36]: It's available in TF as well.

**Review by patil-suraj — APPROVED** [review]
LGTM!

File: src/transformers/pipelines/zero_shot_image_classification.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (patil-suraj) [reply src/transformers/pipelines/zero_shot_image_classification.py:80]: > but it does enable pipe = pipeline(..., candidate_labels["hot-dog", "pizza"]), pipe(image_dataset))

I see. Thanks for the explanation. Then in this case I would also prefer the first option.

File: src/transformers/pipelines/__init__.py (hunk 10 above)
Comment (Narsil) [reply src/transformers/pipelines/__init__.py]: Added support.

**Comment by Narsil** [comment]
@LysandreJik Can you do a second quick review please ? I think adding new pipeline merits a bit more eyes than 4.

File: src/transformers/pipelines/zero_shot_image_classification.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        is_last = model_inputs.pop("is_last")
+        candidate_label = model_inputs.pop("candidate_label")
+        outputs = self.model(**model_inputs)
+
+        # Clip does crossproduct scoring by default, so we're only
+        # interested in the results where image and text and in the same
+        # batch position.
+        diag = torch.diagonal if self.framework == "pt" else tf.linalg.diag_part
+        logits_per_image = diag(outputs.logits_per_image)
+
```
Comment (patil-suraj) [inline src/transformers/pipelines/zero_shot_image_classification.py:106]: Thanks for adding TF support. LGTM!

**Comment by Narsil** [comment]
@patil-suraj Do you think we can add a sigmoid to get `multi_label` or are the outputs of the model not compatible with this ?

**Comment by patil-suraj** [comment]
> @patil-suraj Do you think we can add a sigmoid to get `multi_label` or are the outputs of the model not compatible with this ? @FrancescoSaverioZuppichini

Technically, yes. But I don't know how well that will work.

**Comment by Narsil** [comment]
Ok let's drop it then. I actually thought  about it with maybe multiple prompts too (This photo is about ..., This photo is not about ...) to recover somehow the entailment thing, but CLIP was not trained with this in mind so let's just skip it.)

**Comment by Narsil** [comment]
@LysandreJik friendly ping to get a third opinion before merging.

**Comment by FrancescoSaverioZuppichini** [comment]
After thinking about it, sigmoid won't probably work well since it wasn't trained directly with it. We could (in theory) normalize the `image_logits` and return the ones that are more close to biggest other (meaning they all "fit" the image in the same way). Following @patil-suraj comment, I'm not sure how well this works either.

**Comment by mishig25** [comment]
time to continue the widget PR? https://github.com/huggingface/transformers/pull/12119 😊
